Mathlib/Tactic/Linter/CommandStart.lean

```diff
@@ -0,0 +1,347 @@
+-/
+def isOutside (rgs : Std.HashSet String.Range) (rg : String.Range) : Bool :=
+  rgs.all fun {start := a, stop := b} ↦ !(a ≤ rg.start && rg.stop ≤ b)
+
+/-- `mkWindow orig start ctx` extracts from `orig` a string that starts at the first
+non-whitespace character before `start`, then expands to cover `ctx` more characters
+and continues still until the first non-whitespace character.
+
+In essence, it extracts the substring of `orig` that begins at `start`, continues for `ctx`
+characters plus expands left and right until it encounters the first whitespace character,
+to avoid cutting into "words".
+
+*Note*. `start` is the number of characters *from the right* where our focus is!
+-/
+def mkWindow (orig : String) (start ctx : Nat) : String :=
+  let head := orig.dropRight (start + 1) -- `orig`, up to one character before the discrepancy
+  let middle := orig.takeRight (start + 1)
+  let headCtx := head.takeRightWhile (!·.isWhitespace)
+  let tail := middle.drop ctx |>.takeWhile (!·.isWhitespace)
+  s!"{headCtx}{middle.take ctx}{tail}"
+
+@[inherit_doc Mathlib.Linter.linter.style.commandStart]
+def commandStartLinter : Linter where run := withSetOptionIn fun stx ↦ do
+  unless Linter.getLinterValue linter.style.commandStart (← getLinterOptions) do
+    return
+  if (← get).messages.hasErrors then
+    return
+  if stx.find? (·.isOfKind ``runCmd) |>.isSome then
+    return
+  -- If a command does not start on the first column, emit a warning.
+  if let some pos := stx.getPos? then
+    let colStart := ((← getFileMap).toPosition pos).column
+    if colStart ≠ 0 then
+      Linter.logLint linter.style.commandStart stx
+        m!"'{stx}' starts on column {colStart}, \
+          but all commands should start at the beginning of the line."
+  -- We skip `macro_rules`, since they cause parsing issues.
+  if stx.find? (·.isOfKind `Lean.Parser.Command.macro_rules) |>.isSome then
+    return
+  let some upTo := CommandStart.endPos stx | return
+
+  let fmt : Option Format := ←
+      try
+        liftCoreM <| PrettyPrinter.ppCategory `command stx
+      catch _ =>
+        Linter.logLintIf linter.style.commandStart.verbose (stx.getHead?.getD stx)
+          m!"The `commandStart` linter had some parsing issues: \
+            feel free to silence it and report this error!"
+        return none
+  if let some fmt := fmt then
+    let st := fmt.pretty
+    let origSubstring := stx.getSubstring?.getD default
+    let orig := origSubstring.toString
+
+    let scan := parallelScan orig st
+
+    let docStringEnd := stx.find? (·.isOfKind ``Parser.Command.docComment) |>.getD default
+    let docStringEnd := docStringEnd.getTailPos? |>.getD default
+    let forbidden := getUnlintedRanges unlintedNodes ∅ stx
+    for s in scan do
+      let center := origSubstring.stopPos - s.srcEndPos
+      let rg : String.Range := ⟨center, center + s.srcEndPos - s.srcStartPos + ⟨1⟩⟩
+      if s.msg.startsWith "Oh no" then
```

Comment:
So it looks like we are doing a bit of pattern-matching with the message :) I'll update my suggestion above.